### PR TITLE
✨ Add e2e tests scaffolded in default layout

### LINKED ISSUE #3605: Add e2e tests scaffolded in the default layout
### What do you want to happen?

The goal of this task is scaffold in the default layout e2e tests such as: https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/memcached-operator/test

The implementation should be:
- a) For go/v4 code implementation: we need to cover the Operator itself 
- b) For deploy-image plugin we need increment the tests to add more checks regards the deployment scaffolded by the plugin 

Also, we must ensure that we check the e2e tests with the ci, so we not allow get PR merges that broken then. To do so we can run an ci which will call the tests scaffold for project-v4 and project-v4-with-deploy-image under the testdata. 

**Motivation** Users does not know how to implement e2e tests in the projects. Also, provide to them it done since it is possible. The e2e tests can be configured to run against kind since we need to load the image. We also should add a TODO(user) when we load the image in the kind to explain to users that they need to change that if they do not want to run the tests using kind. 

Example motivation: https://kubernetes.slack.com/archives/CAW0GV7A5/p1694137817036729

### Extra Labels

_No response_

### PR DESCRIPTION
**Pull request by Lauquik**
## Description

adding testing scaffold for the default layout of our project

Test perfoms the follwoing tasks
- Building the Operator Image:
- Loading the Operator Image to Kind:
- Deploying the Controller-Manager:
- Validating Controller-Manager Pod:
## Fixes
#3605

### CODE REVIEW

File: pkg/plugins/golang/v4/scaffolds/internal/templates/test/utils/utils.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+// Run executes the provided command within this context
+func Run(cmd *exec.Cmd) ([]byte, error) {
+	dir, _ := GetProjectDir()
+	cmd.Dir = dir
+	fmt.Fprintf(GinkgoWriter, "running dir: %s\n", cmd.Dir)
+
+	// To allow make commands be executed from the project directory which is subdir on SDK repo
+	// TODO:(user) You might not need the following code
```
Comment (camilamacedo86) [inline pkg/plugins/golang/v4/scaffolds/internal/templates/test/utils/utils.go:62]: It should be removed the end project does not need it at all.
We just have each in the operator-sdk project to be able to run the tests due how that is configured there.

File: test/e2e/ci.sh [hunk 2]
```diff
@@ -24,3 +24,13 @@ create_cluster ${KIND_K8S_VERSION}
...
+
+KUSTOMIZATION_FILE_PATH="$(dirname "$0")/../../testdata/project-v4/config/default/kustomization.yaml"
+
+sed -i '25s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '27s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '42s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '46,143s/^#//' $KUSTOMIZATION_FILE_PATH
+
+cd $(dirname "$0")/../../testdata/project-v4
+make test-e2e
```
Comment (camilamacedo86) [inline test/e2e/ci.sh:36]: This one is called to do the prow tests
I think we need a new configuration to perform this specific tests
Ideally, use GitHub Actions instead. 
So that would result in a new GitHub action that will setup the kind and run the tests scaffold for the sample

File: test/testdata/test.sh [hunk 3]
```diff
@@ -23,6 +23,7 @@ function test_project {
 
   header_text "Performing tests in dir $1"
   pushd "$(dirname "$0")/../../testdata/$1"
+  rm -rf test/
```
Comment (camilamacedo86) [inline test/testdata/test.sh:26]: This check ensures that the testdata samples are updated with the latest changes.
Why are we removing tests/ dir from here?

File: testdata/project-v4-multigroup-with-deploy-image/api/crew/v1/webhook_suite_test.go [hunk 4]
```diff
@@ -73,7 +73,7 @@ var _ = BeforeSuite(func() {
 		// Note that you must have the required binaries setup under the bin directory to perform
 		// the tests directly. When we run make test it will be setup and used automatically.
 		BinaryAssetsDirectory: filepath.Join("..", "..", "..", "bin", "k8s",
-			fmt.Sprintf("1.28.3-%s-%s", runtime.GOOS, runtime.GOARCH)),
+			fmt.Sprintf("1.28.0-%s-%s", runtime.GOOS, runtime.GOARCH)),
```
Comment (camilamacedo86) [inline testdata/project-v4-multigroup-with-deploy-image/api/crew/v1/webhook_suite_test.go:76]: You need rebase your PR with master changes.

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
Great work, but I think we still require some adjusts
- Rebase with master 
- Do not change the current e2e tests to try run the sample tests. The e2e tests executed run the e2e tests scaffolds for kubebuilder itself. We need a new GitHub action that will run the sample tests and be configured with kind
- Ensure that the final scaffold has only what is required for end users.

File: test/testdata/test.sh (hunk 3 above)
Comment (Lauquik) [reply test/testdata/test.sh:26]: if the test dir is not removed 
then scaffolded e2e tests will run for all the sample which we don't require right? we only want to test one sapmle(project-v4)

Comment (camilamacedo86) [reply test/testdata/test.sh:26]: The testdata check if all samples are == they should be with the latest changes.
We are changing the default scaffold so all v4 samples should have the tests now.

We do not need to execute and run the e2e tests against all projects
If we see that does not take too long that is fine would be great. But if we see that take too long then we just test 1/2 samples. But that is related to the new GitHub action and not the change test data check.

is that make sense?

File: test/e2e/sample-e2e.sh [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+KUSTOMIZATION_FILE_PATH="$(dirname "$0")/../../testdata/project-v4/config/default/kustomization.yaml"
+
+sed -i '25s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '27s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '42s/^#//' $KUSTOMIZATION_FILE_PATH
+sed -i '46,143s/^#//' $KUSTOMIZATION_FILE_PATH
+
+cd $(dirname "$0")/../../testdata/project-v4
+go get -u ./...
+make test-e2e
```
Comment (camilamacedo86) [inline test/e2e/sample-e2e.sh:35]: Nit: add a space at the end of the file.

File: test/e2e/ci.sh [hunk 6]
```diff
@@ -23,4 +23,4 @@ export KIND_CLUSTER="kind"
 create_cluster ${KIND_K8S_VERSION}
 trap delete_cluster EXIT
 
-test_cluster
+test_cluster
```
Comment (camilamacedo86) [inline test/e2e/ci.sh:26]: Nit: add a space at the end of the file.

File: .github/workflows/test-sample-go.yml [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+# Trigger the workflow on pull requests and direct pushes to any branch
+on:
+  push:
+  pull_request:
+
+jobs:
+  test:
+    name: ${{ matrix.os }}
+    runs-on: ${{ matrix.os }}
```
Comment (camilamacedo86) [inline .github/workflows/test-sample-go.yml:11]: We do not need a matrix.os here
We just do the test using ubuntu linux for this one.

File: Makefile [hunk 8]
```diff
@@ -152,10 +152,15 @@ test-e2e-ci: ## Run the end-to-end tests (used in the CI)`
...
+	cd ./docs/book/src/multiversion-tutorial/testdata/project && rm -rf test && make test
 
 .PHONY: test-license
 test-license:  ## Run the license check
 	./test/check-license.sh
+
+.PHONY: test-e2e-sample
+test-e2e-sample:
+	./test/e2e/sample-e2e.sh
+
```
Comment (camilamacedo86) [inline Makefile:166]: Nit extra space

File: test/e2e/sample-e2e.sh [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+source "$(dirname "$0")/../common.sh"
+source "$(dirname "$0")/setup.sh"
+
+build_sample_external_plugin
+
+export KIND_CLUSTER="kind"
```
Comment (camilamacedo86) [inline test/e2e/sample-e2e.sh:22]: I do not think that will work
Instead of create a new sample-e2e.sh
You can use github actions to install the kind
Then you go into the sample dir and run go mod tidy and make, and by last, the command make test-e2e to run the tests against kind. 

is that make sense?

**Comment by Lauquik** [comment]
@camilamacedo86 i have created a whole new workflow for sample project-v4 and removed the sample-e2e.sh script, is it okay? or require any changes, i'm open to receiving any suggestion and guidance

**Comment by camilamacedo86** [comment]
Hi @Lauquik,

> @camilamacedo86 i have created a whole new workflow for sample project-v4 and removed the sample-e2e.sh script, is it okay? or require any changes, i'm open to receiving any suggestion and guidance

- The samples should continue to be generated as they currently are.
- The existing checks should remain operational in their current form.

That said, to ensure the integrity of our new e2e tests, we need a dedicated GitHub Action. While it's not necessary to test every single sample (and if doing so simplifies the process, that's acceptable), a streamlined approach might be:

1. Set up a new GitHub Action.
2. Have this action install kind.
3. Run the e2e tests that have been scaffolded for example for project-v4.

File: .github/workflows/test-sample-go.yml [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+          sed -i '27s/^#//' $KUSTOMIZATION_FILE_PATH
+          sed -i '42s/^#//' $KUSTOMIZATION_FILE_PATH
+          sed -i '46,143s/^#//' $KUSTOMIZATION_FILE_PATH
+      
+      - name: Test
+        run: |
+          cd testdata/project-v4
+          go get -u ./...
+          go mod tidy
+          make test-e2e
```
Comment (camilamacedo86) [inline .github/workflows/test-sample-go.yml:50]: Nit: missing space

File: Makefile [hunk 11]
```diff
@@ -152,9 +152,9 @@ test-e2e-ci: ## Run the end-to-end tests (used in the CI)`
 
 .PHONY: test-book
 test-book: ## Run the cronjob tutorial's unit tests to make sure we don't break it
-	cd ./docs/book/src/cronjob-tutorial/testdata/project && make test
-	cd ./docs/book/src/component-config-tutorial/testdata/project && make test
-	cd ./docs/book/src/multiversion-tutorial/testdata/project && make test
+	cd ./docs/book/src/cronjob-tutorial/testdata/project && rm -rf test && make test
+	cd ./docs/book/src/component-config-tutorial/testdata/project && rm -rf test && make test
+	cd ./docs/book/src/multiversion-tutorial/testdata/project && rm -rf test && make test
```
Comment (camilamacedo86) [inline Makefile:157]: Why did you remove the tests from here?
Is the make test calling the new e2e tests?
If so, the best solution is not call the e2e tests in the make test

File: .github/workflows/lint-sample.yml [hunk 12]
```diff
@@ -21,9 +21,9 @@ jobs:
       - name: Remove pre-installed kustomize
         run: sudo rm -f /usr/local/bin/kustomize
       - name: Run make test for project-v4-with-deploy-image
-        run: cd testdata/project-v4-with-deploy-image && go mod tidy && make all && make test
+        run: cd testdata/project-v4-with-deploy-image && rm -rf test && go mod tidy && make all && make test
```
Comment (camilamacedo86) [inline .github/workflows/lint-sample.yml:24]: same here, we should not remove the tests from the lint

**Review by camilamacedo86 — COMMENTED** [review]
Hi @Lauquik 

Very terrific great work 🥇 
I think we have just a couple of nits to sort out, and then it is good to fly.

File: test/testdata/test.sh (hunk 3 above)
Comment (camilamacedo86) [reply test/testdata/test.sh:26]: Why the e2e tests will run for all samples?
We are calling make test, right?

Also, see that it is not a problem to run the e2e tests for all samples (unless it would take too long)
The idea to run for one for then is to simplify the implementation.

File: Makefile (hunk 11 above)
Comment (Lauquik) [reply Makefile:157]: yeah, whenever make test is called (https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/Makefile#L65 )
at the end of the line there is go test ./... which will run the go test in every directory including the e2e dir where test suite is present

**Comment by Lauquik** [comment]
@camilamacedo86 i think you are right removing the test dir is not the best approach, i will think of some other solution

**Comment by Lauquik** [comment]
@camilamacedo86  I have modified the scaffolded Makefile to exclude the e2e test when running 'make test,' so there's no need to remove the test directories explicitly

**Review by camilamacedo86 — APPROVED** [review]
Great work 🥇 
It is great for me so I am approving this one

/approved

Let's give some time to see if others also want to do some review.
If nobody came with a second review or objections we can lgtm this one.

**Review by Kavinjsir — APPROVED** [review]
Great job @Lauquik 👍🏼  The PR Looks good to me!

**Review by camilamacedo86 — APPROVED** [review]
/lgtm

It seems good to fly
we can always push improvements and etc in a follow up
